scrapy_autoextract/providers.py

```diff
@@ -116,13 +122,30 @@ async def create_aiohttp_session(self) -> aiohttp.ClientSession:
         )
         return create_session(connection_pool_size=concurrent_connections)
 
-    async def close_aiohttp_session(self):
+    async def on_spider_closed(self):
         if self.aiohttp_session:
             await self.aiohttp_session.close()
+        self.cache.close()
 
     def create_retry_wrapper(self):
         return RetryFactory().build()
 
+    async def do_request_cached(self, query, *args, **kwargs):
+        assert len(query) == 1  # batches are not supported
+        fp = self.cache.fingerprint(query[0])
+        try:
+            response = self.cache[fp]
+            self.stats.inc_value("autoextract/cache/hit")
+        except KeyError:
+            self.stats.inc_value("autoextract/cache/miss")
+            response = await self.do_request(query, *args, **kwargs)
+            if "error" not in response[0]:  # don't cache errors
```

Comment:
Yeah, I was unsure about it as well. There are use cases for both behaviors. We should probably have an option for this. 

My motivation for going "don't cache errors" by default was this: you've got an error, and most likely it is some proxy error. It may resolve on itself, but it may also be resolved by contacting support. This could be a part of iterative development, a standard one - downloading doesn't work or work unreliably => you contact support => error is fixed. And in these cases you'd need to remove the cache and re-create it to get fixes. The question is, do you get useful data if such situation happens, or is it all errors? if that's all errors, than not caching them doesn't help.

You may also have a goal of producing a dataset. In this case, if there were some errors, you just rerun the spider, and if errors are resolved (either on their own, or by contacting support), you get an updated dataset quickly. 

But I think ultimately practice would tell what's better. As this is mostly a development tool, I think that's fine to change the default behavior later if we want to.

I'm fine with changing it, or adding an option. Your call @ivanprado - what should the initial version be? :)

Reply:
The first usage of this tool will benefit from the current implementation, so let's merge it as is by now.